### PR TITLE
fix: Rewrite time range expressions so they apply to entire result set

### LINKED ISSUE #7610: InfluxQL-IOx: Time range semantics should match InfluxQL OG
The time range predicate must be applied to the entire query such that

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE cpu = 'cpu0' OR cpu = 'cpu1' AND time > '2020-06-11T16:53:30Z'
```

is the equivalent of

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE (cpu = 'cpu0' OR cpu = 'cpu1') AND time > '2020-06-11T16:53:30Z'
```

and they must be rewritten using conjunction (`OR`) operators, regardless of how they appear in the original expression.

## Background

The InfluxQL query engine always promotes the time range expression to filter all results. It is misleading that time ranges are written in the `WHERE` clause, as the `WHERE` predicate is not evaluated in its entirety for each row. Rather, InfluxQL extracts the time range to form a time bound for the entire query and removes any time range expressions from the filter predicate. The time range is determined using the `>` and `≥` operators to form the lower bound and the `<` and `≤` operators to form the upper bound. When multiple instances of the lower or upper bound operators are found, the time bounds will form the intersection. For example

```sql
WHERE time >= 1000 AND time >= 2000 AND time < 10000 and time < 9000
```

is equivalent to

```sql
WHERE time >= 2000 AND time < 9000
```

Further, InfluxQL only allows a single `time = <value>` binary expression. Multiple occurrences result in an empty result set.

### Examples

Lets illustrate how InfluxQL applies predicates with a typical example, using the `metrics.lp` data in the IOx repository:

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE 
  time > '2020-06-11T16:53:30Z' AND time < '2020-06-11T16:55:00Z' AND cpu = 'cpu0'
```

InfluxQL first filters rows based on the time range: 

```
'2020-06-11T16:53:30Z' < time <  '2020-06-11T16:55:00Z'
```

and then applies the predicate to the individual rows:

```
cpu = 'cpu0'
```

Producing the following result:

```text
name: cpu
time                 cpu  usage_idle
----                 ---  ----------
2020-06-11T16:53:40Z cpu0 90.29029029029029
2020-06-11T16:53:50Z cpu0 89.8
2020-06-11T16:54:00Z cpu0 90.09009009009009
2020-06-11T16:54:10Z cpu0 88.82235528942115
```

The following example is a little more complicated, but shows again how InfluxQL separates the time ranges from the predicate:

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE 
  time > '2020-06-11T16:53:30Z' AND time < '2020-06-11T16:55:00Z' AND cpu = 'cpu0' OR cpu = 'cpu1'
```

InfluxQL first filters rows based on the time range: 

```
'2020-06-11T16:53:30Z' < time <  '2020-06-11T16:55:00Z'
```

and then applies the predicate to the individual rows:

```
cpu = 'cpu0' OR cpu = 'cpu1'
```

This is certainly quite different to SQL, which would evaluate the predicate as:

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE 
  (time > '2020-06-11T16:53:30Z' AND time < '2020-06-11T16:55:00Z' AND cpu = 'cpu0') OR cpu = 'cpu1'
```

#### Time ranges are not normal

Finally, we demonstrate how the operators combining time ranges do not matter. Lets take the original query:

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE 
  time > '2020-06-11T16:53:30Z' AND time < '2020-06-11T16:55:00Z' AND cpu = 'cpu0'
```

and rewrite it with `OR` operators

```sql
SELECT cpu, usage_idle FROM cpu 
WHERE 
  time > '2020-06-11T16:53:30Z' OR time < '2020-06-11T16:55:00Z' OR cpu = 'cpu0'
```

This should return all rows, but yet it returns the same result 🤯:

```text
name: cpu
time                 cpu  usage_idle
----                 ---  ----------
2020-06-11T16:53:40Z cpu0 90.29029029029029
2020-06-11T16:53:50Z cpu0 89.8
2020-06-11T16:54:00Z cpu0 90.09009009009009
2020-06-11T16:54:10Z cpu0 88.82235528942115
```

It becomes clearer, if we again review at how InfluxQL OG evaluates the `WHERE` predicate, InfluxQL first filters rows based on the time range, which uses the rules previously defined by finding `>` and `≥` to determine the lower bound and `<` and `≤`: 

```
'2020-06-11T16:53:30Z' < time <  '2020-06-11T16:55:00Z'
```

and then applies the predicate to the individual rows:

```
cpu = 'cpu0'
```

### How to think of it intuitively

Imagine a slight variation of InfluxQL has a separate _time bounds clause_. It could have two forms, first as a `BETWEEN`

```sql
SELECT cpu, usage_idle FROM cpu 
WITH TIME BETWEEN '2020-06-11T16:53:30Z' AND '2020-06-11T16:55:00Z'
WHERE 
  cpu = 'cpu0'
```

or as an `IN` to select multiple points

```sql
SELECT cpu, usage_idle FROM cpu 
WITH TIME IN ('2004-04-09T12:00:00Z', '2004-04-09T12:00:10Z', ...)
WHERE 
  cpu = 'cpu0'
```

## Problem

Currently InfluxQL on IOx is passing the entire filter predicate to DataFusion, meaning the following query:

```sql
select cpu, usage_idle from cpu where cpu = 'cpu0' or cpu = 'cpu1' and time > '2020-06-11T16:53:30Z'
```

is equivalent to:

```sql
select cpu, usage_idle from cpu where cpu = 'cpu0' or (cpu = 'cpu1' and time > '2020-06-11T16:53:30Z')
```

and therefore returning incorrect results with respect to InfluxQL OG. We can validate this by wrapping the `OR` condition in parenthesis to produce the same results as InfluxQL OSS:

```sql
select cpu, usage_idle from cpu where (cpu = 'cpu0' or cpu = 'cpu1') and time > '2020-06-11T16:53:30Z'
```

## Steps

Using `metrics.lp` in the IOx repository, execute

1. Execute

    ```sql
    select cpu, usage_idle from cpu where cpu = 'cpu0' or cpu = 'cpu1' and time > '2020-06-11T16:53:30Z'
    ```

## Expected results

time > '2020-06-11T16:53:30Z' is applied to entire result set producing:

```text
name: cpu
time                 cpu  usage_idle
----                 ---  ----------
2020-06-11T16:53:40Z cpu0 90.29029029029029
2020-06-11T16:53:40Z cpu1 99.89989989989989
2020-06-11T16:53:50Z cpu0 89.8
2020-06-11T16:53:50Z cpu1 99.8001998001998
2020-06-11T16:54:00Z cpu0 90.09009009009009
2020-06-11T16:54:00Z cpu1 99.8997995991984
2020-06-11T16:54:10Z cpu0 88.82235528942115
2020-06-11T16:54:10Z cpu1 99.8001998001998
```

## Actual results

```text
name: cpu
time                 cpu  usage_idle
----                 ---  ----------
2020-06-11T16:52:00Z cpu0 89.56262425447316
2020-06-11T16:52:10Z cpu0 88.6113886113886
2020-06-11T16:52:20Z cpu0 88.6
2020-06-11T16:52:30Z cpu0 83.41708542713567
2020-06-11T16:52:40Z cpu0 87.76119402985074
2020-06-11T16:52:50Z cpu0 88.78878878878879
2020-06-11T16:53:00Z cpu0 86.91308691308691
2020-06-11T16:53:10Z cpu0 89.88988988988989
2020-06-11T16:53:20Z cpu0 89.01098901098901
2020-06-11T16:53:30Z cpu0 90.4
2020-06-11T16:53:40Z cpu0 90.29029029029029
2020-06-11T16:53:40Z cpu1 99.89989989989989
2020-06-11T16:53:50Z cpu0 89.8
2020-06-11T16:53:50Z cpu1 99.8001998001998
2020-06-11T16:54:00Z cpu0 90.09009009009009
2020-06-11T16:54:00Z cpu1 99.8997995991984
2020-06-11T16:54:10Z cpu0 88.82235528942115
2020-06-11T16:54:10Z cpu1 99.8001998001998
```

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #7610

This PR ensures the semantics of time range expressions behave as described in #7160

### CODE REVIEW

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs [hunk 1]
```diff
@@ -392,11 +594,64 @@ fn rewrite_boolean(lhs: Expr, op: Operator, rhs: Expr) -> Expr {
     }
 }
 
+/// Rewrite regex conditional expressions to match InfluxQL behaviour.
```
Comment (wolffcm) [inline iox_query_influxql/src/plan/planner_rewrite_expression.rs:597]: I'm assuming that this was not changed, and just moved from its old location.

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs [hunk 2]
```diff
@@ -121,16 +121,218 @@
...
+        }
+    })?;
+
+    if !has_or {
+        return Ok(expr);
+    }
+
+    let mut rw = SeparateTimeRanges {
+        ..Default::default()
+    };
```
Comment (wolffcm) [inline iox_query_influxql/src/plan/planner_rewrite_expression.rs:189]: This type derives `Default` so seems like you could just do
```suggestion
    let mut rw = SeparateTimeRanges::default();
```

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs [hunk 3]
```diff
@@ -121,16 +121,218 @@
...
+///
+/// ```text
+/// time > now() - 5m
+/// ```
+fn is_time_range(expr: &Expr) -> bool {
+    use Operator::*;
+    match expr {
+        Expr::BinaryExpr(BinaryExpr {
+            left,
+            op: Eq | NotEq | Gt | Lt | GtEq | LtEq,
```
Comment (wolffcm) [inline iox_query_influxql/src/plan/planner_rewrite_expression.rs:245]: I am still surprised sometimes at how expressive Rust patterns can be.

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs [hunk 4]
```diff
@@ -121,16 +121,218 @@
...
+                    .ok_or_else(|| error::map::internal("invalid expr stack"))?;
+                let left = self
+                    .stack
+                    .pop()
+                    .ok_or_else(|| error::map::internal("invalid expr stack"))?;
+                self.stack.push(match (left, right) {
+                    (Some(left), Some(right)) => Some(binary_expr(left, *op, right)),
+                    (None, Some(node)) | (Some(node), None) => Some(node),
+                    (None, None) => None,
+                });
```
Comment (wolffcm) [inline iox_query_influxql/src/plan/planner_rewrite_expression.rs:325]: Wow, ok. So this seems to be treating
```
time > now() - 5m  AND tag1 = 'foo'
```
and
```
time > now() - 5m OR tag1 = 'foo'
```
Exactly the same?

**Review by wolffcm — APPROVED** [review]
This seems reasonable to me. Thank you!

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs (hunk 4 above)
Comment (wolffcm) [reply iox_query_influxql/src/plan/planner_rewrite_expression.rs:325]: I thought I had mentioned this somewhere but I don't see it now. A lot of this behavior seems to have come about in 1.x only by accident, and is not intuitive. Especially operator precedence for AND and OR not working like one would expect seems strange.

I understand why we want to preserve the OG behavior, especially for this case. Maybe at some point we can fix it.

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs (hunk 1 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner_rewrite_expression.rs:597]: You are correct 👍🏻

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs (hunk 4 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner_rewrite_expression.rs:323]: Yeah, you are exactly right. I expanding the original issue to describe the behaviour. I am going to copy the **Background** section to this file so it is clear to future readers.

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs (hunk 3 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner_rewrite_expression.rs:399]: Me too, you can even bind all those use:

```rust
op: op @ (Eq | NotEq | Gt | Lt | GtEq | LtEq)
```

so you can use `op` in the body of the match arm

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs (hunk 4 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner_rewrite_expression.rs:479]: @wolffcm, I've documented the entire behaviour of InfluxQL time ranges:

https://github.com/influxdata/influxdb_iox/blob/7a49c3484eae724a1841a0e5330bc9251bb097bc/iox_query_influxql/src/plan/planner_rewrite_expression.rs#L165-L319

Comment (wolffcm) [reply iox_query_influxql/src/plan/planner_rewrite_expression.rs:479]: That is super clear, thanks for taking the time.

**Pull request merged**